test/sql-luatest/constraint_test.lua

```diff
@@ -609,3 +609,88 @@ g.test_drop_constraints_with_the_same_name = function()
         box.func.check_T_C:drop()
     end)
 end
+
+-- Make sure the variations of the DROP CONSTRAINT statement with the specified
+-- constraint type work correctly.
+g.test_drop_constraints_with_type = function()
+    g.server:exec(function()
+        local sql = [[CREATE TABLE t (i INT CONSTRAINT c PRIMARY KEY,
+                                      a INT CONSTRAINT b CHECK (a > 10)
+                                      CONSTRAINT b REFERENCES t(i),
+                                      CONSTRAINT c CHECK (i + a > 100),
+                                      CONSTRAINT c FOREIGN KEY (a) REFERENCES t,
+                                      CONSTRAINT d UNIQUE(a));]]
+        box.execute(sql)
+        local s = box.space.T
+        t.assert(s.constraint['C'] ~= nil);
+        t.assert(s.foreign_key['C'] ~= nil);
+        t.assert(s.index['C'] ~= nil);
+        t.assert(s.index['D'] ~= nil);
+        t.assert(s:format()[2].constraint['B'] ~= nil);
+        t.assert(s:format()[2].foreign_key['B'] ~= nil);
+
+        -- Make sure that a constraint with the given name but with the wrong
+        -- type will not be dropped.
+        local _, err = box.execute([[ALTER TABLE t DROP CONSTRAINT d CHECK;]])
+        local exp = "Constraint 'D' does not exist in space 'T'"
```

Comment:
Well, the error is quite misguiding: the constraint with the name 'D' exists but its type is UNIQUE, not CHECK. It would be nice to mention it in the error, but if it requires a complex fix in the parser, I suggest to omit such a change in scope of this PR, but file the issue describing the problem.

Reply:
This does not require much, but I think it would be better to leave it as it is for now. There is quite a few errors that have to be reworked, and I plan to do it later.

Reply:
OK, that works for me.